### ISSUE TITLE
AMSI logging implemented on Linux

### ISSUE THREAD

**Issue by rhubarb-geek-nz** [PowerShell/PowerShell#21492]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

The AMSI logging is still being performed on Linux even though there are no AMSI modules installed.

As demonstrated in [#21473](https://github.com/PowerShell/PowerShell/issues/21473) this can lead to excessive heap usage

### Expected behavior

```console
#!/usr/bin/env pwsh
$env:__PSDumpAMSILogContent='1'
$bytes = new-object byte[] -ArgumentList @(,200554320)
$random = new-object Random
$random.NextBytes($bytes)

Should not print anything
```

### Actual behavior

```console
#!/usr/bin/env pwsh
$env:__PSDumpAMSILogContent='1'
$bytes = new-object byte[] -ArgumentList @(,200554320)
$random = new-object Random
$random.NextBytes($bytes)

Prints the following log

=== Amsi notification report content ===
<System.Random>.NextBytes(<System.Byte[]>)
=== Amsi notification report success: False ===
```

### Error details

```console
No actual error other than incorrect behaviour
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.2
PSEdition                      Core
GitCommitId                    7.4.2
OS                             Debian GNU/Linux 11 (bullseye)
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iSazonov** [comment]
AMSI is called in `internal static bool ReportContent()` and only on Windows.
What you see is a debug code. It's not worth the effort to mask it on Unix. (Maybe it can even be useful there for MSFT developers who approved the code.)

**Comment by rhubarb-geek-nz** [comment]
> What you see is a debug code. It's not worth the effort to mask it on Unix. 

What the system is doing for every .NET invocation on Linux and macOS is linearising all the string arguments then throwing the result away. For _every_ .NET invocation.

**Comment by mklement0** [comment]
@iSazonov, I agree with @rhubarb-geek-nz that there's no good reason to saddle user on Unix-like platform not only with unnecessary _runtime_ overhead, but - as noted - in the case of passing large strings also with significant _memory_ overhead.

It would be simple to execute the code in question _only if env var. `DumpLogAMSIContent` is set to `1`_ on Unix, which would preserve the ability to debug _on demand_, while not burdening regular method calls:

That is, the following code:

https://github.com/PowerShell/PowerShell/blob/c19b65133c00633c28af0fa0a75b2c559de9d0ee/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L6938-L6950

could be guarded by a  preceding `#if UNIX` directive that sets a Boolean to determine whether the code should be invoked, along the lines of:

```csharp
#if UNIX
  bool callAMSI = Environment.GetEnvironmentVariable("__PSDumpAMSILogContent").Trim() == "1";
#else
  bool callAMSI = true;
#endif
if (callAMSI)
{
   // ... code above
}
```

Note: The above illustrates the concept, but the value of `callAMSI` can and should be determined only once per session and then cached.

**Comment by iSazonov** [comment]
For performance there is another issue. You can ask MSFT team today on monthly community call if you want to get answer quickly about #21473 performance.

**Comment by mklement0** [comment]
@iSazonov:

* The memory usage problem and the - much worse - runtime penalty discussed in #21473 is in the context of _Windows_, where actual AMSI API calls are currently performed _unconditionally_.

  * As long as an AMSI call must be made, these memory + performance issues are unavoidable, although the only really _pathological_ cases are .NET method calls involving arguments that are _very large strings_.

  * Possibly providing an opt-out of these AMSI calls on Windows - something that obviously has security consideration - is the subject of yet another issue, #21491

* The issue at hand is purely about _Unix_-like platforms, where no security considerations apply, due to the absence of an API analogous to the Windows-only AMSI, so this issue can and should be decided separately.

**Comment by mklement0** [comment]
@iSazonov, in light of the above:

* The `Resolution-Answered` label is inappropriate.

* While it is appropriate to make the security WG review this, let me summarize what I think the net effect of this proposal is:
  * It will result in _no functional change_, and therefore _doesn't affect security_.
  * It will generally improve performance of all .NET method calls on all Unix-like platforms.

**Comment by rhubarb-geek-nz** [comment]
Although this was marked as answered, it was neither a solution or an acknowledgment that there is any kind of problem.

**Comment by SteveL-MSFT** [comment]
WG reviewed this and agreed that the .NET method invocation logging for AMSI should only apply to Windows

**Comment by SteveL-MSFT** [comment]
I made the change, but now noticed an unexpected behavior.  With the `Console.WriteLine()` calls now `Console.Error.WriteLine()`, you don't get ANY output in the interactive shell.  STDERR is only written if you call `pwsh` itself.  It might make more sense to just remove this env var altogether as it's not useful expect as a test hook.

**Comment by jborean93** [comment]
Opened https://github.com/PowerShell/PowerShell/pull/24451 which no-ops the member invocation logging for non-Windows on a release build. Keep in mind the extra overhead added by serialising the args is pretty minimal but always good to avoid it when possible.

**Comment by iSazonov** [comment]
There is a bug in https://github.com/PowerShell/PowerShell/pull/24451.

**Comment by jborean93** [comment]
@iSazonov bug has been fixed with https://github.com/PowerShell/PowerShell/pull/24652
